From a3e4b399b4daced8e5e3f4f0fbf081b1a1d4f8c4 Mon Sep 17 00:00:00 2001
From: Dinesh Sudham <dineshx.sudham@intel.com>
Date: Tue, 4 Aug 2020 11:41:58 +0800
Subject: [PATCH] DRVLIB_SW-3774:GSWAPI: use spinlock_bh when data shares
 between user and bottom halves context

current context can be interrupted by a bottom half and critical
region can be excuted from another CPU, use spinlock_bh it disables
bottom halves on local CPU and get lock.

Signed-off-by: Dinesh Sudham <dineshx.sudham@intel.com>
---
 .../net/ethernet/lantiq/switch-api/gsw_tbl_rw.c    | 24 +++++++++++-----------
 1 file changed, 12 insertions(+), 12 deletions(-)

diff --git a/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c b/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c
index fc38d7f76409..f2bb6f4bd252 100644
--- a/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c
+++ b/drivers/net/ethernet/lantiq/switch-api/gsw_tbl_rw.c
@@ -168,7 +168,7 @@ int route_table_read(void *cdev, pctbl_prog_t *rdata)
 	} while (value != 0);
 
 #ifdef __KERNEL__
-	spin_lock(&gswdev->lock_pce_tbl);
+	spin_lock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	gsw_w32(cdev, PCE_TBL_ADDR_ADDR_OFFSET,
@@ -330,7 +330,7 @@ int route_table_read(void *cdev, pctbl_prog_t *rdata)
 	gsw_w32(cdev, PCE_TBL_CTRL_ADDR_OFFSET, 0, 16, 0);
 
 #ifdef __KERNEL__
-	spin_unlock(&gswdev->lock_pce_tbl);
+	spin_unlock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	return GSW_statusOk;
@@ -349,7 +349,7 @@ int route_table_write(void *cdev, pctbl_prog_t *rdata)
 	} while (value);
 
 #ifdef __KERNEL__
-	spin_lock(&gswdev->lock_pce_tbl);
+	spin_lock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	gsw_w32(cdev, PCE_TBL_ADDR_ADDR_OFFSET,
@@ -482,7 +482,7 @@ int route_table_write(void *cdev, pctbl_prog_t *rdata)
 	gsw_w32(cdev, PCE_TBL_CTRL_ADDR_OFFSET, 0, 16, 0);
 
 #ifdef __KERNEL__
-	spin_unlock(&gswdev->lock_pce_tbl);
+	spin_unlock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	return GSW_statusOk;
@@ -505,7 +505,7 @@ int gsw_pce_table_write(void *cdev, pctbl_prog_t *ptdata)
 			       PCE_TBL_CTRL_BAS_SIZE));
 
 #ifdef __KERNEL__
-	spin_lock(&gswdev->lock_pce_tbl);
+	spin_lock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	gsw_w32_raw(cdev, PCE_TBL_ADDR_ADDR_OFFSET, ptdata->pcindex);
@@ -560,7 +560,7 @@ int gsw_pce_table_write(void *cdev, pctbl_prog_t *ptdata)
 	gsw_w32_raw(cdev, PCE_TBL_CTRL_ADDR_OFFSET, 0);
 
 #ifdef __KERNEL__
-	spin_unlock(&gswdev->lock_pce_tbl);
+	spin_unlock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	return GSW_statusOk;
@@ -578,7 +578,7 @@ static int __gsw_pce_table_read(void *cdev, ethsw_api_dev_t *gswdev,
 			       PCE_TBL_CTRL_BAS_SIZE));
 
 #ifdef __KERNEL__
-	spin_lock(&gswdev->lock_pce_tbl);
+	spin_lock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	gsw_w32_raw(cdev, PCE_TBL_ADDR_ADDR_OFFSET, ptdata->pcindex);
@@ -635,7 +635,7 @@ static int __gsw_pce_table_read(void *cdev, ethsw_api_dev_t *gswdev,
 	gsw_w32_raw(cdev, PCE_TBL_CTRL_ADDR_OFFSET, 0);
 
 #ifdef __KERNEL__
-	spin_unlock(&gswdev->lock_pce_tbl);
+	spin_unlock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	return GSW_statusOk;
@@ -697,7 +697,7 @@ int gsw_pce_table_key_read(void *cdev, pctbl_prog_t *ptdata)
 			       PCE_TBL_CTRL_BAS_SIZE));
 
 #ifdef __KERNEL__
-	spin_lock(&gswdev->lock_pce_tbl);
+	spin_lock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	/*KEY REG*/
@@ -742,7 +742,7 @@ int gsw_pce_table_key_read(void *cdev, pctbl_prog_t *ptdata)
 	gsw_w32_raw(cdev, PCE_TBL_CTRL_ADDR_OFFSET, 0);
 
 #ifdef __KERNEL__
-	spin_unlock(&gswdev->lock_pce_tbl);
+	spin_unlock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	return GSW_statusOk;
@@ -766,7 +766,7 @@ int gsw_pce_table_key_write(void *cdev, pctbl_prog_t *ptdata)
 			       PCE_TBL_CTRL_BAS_SIZE));
 
 #ifdef __KERNEL__
-	spin_lock(&gswdev->lock_pce_tbl);
+	spin_lock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	ctrlval = 0;
@@ -818,7 +818,7 @@ int gsw_pce_table_key_write(void *cdev, pctbl_prog_t *ptdata)
 	gsw_w32_raw(cdev, PCE_TBL_CTRL_ADDR_OFFSET, 0);
 
 #ifdef __KERNEL__
-	spin_unlock(&gswdev->lock_pce_tbl);
+	spin_unlock_bh(&gswdev->lock_pce_tbl);
 #endif
 
 	return GSW_statusOk;
